### PR TITLE
Add `Array#delete_at(index, count)` and `Array#delete_at(range)`.

### PR DESCRIPTION
**Pull request by tebakane**
Instead of slice, I expand `delete_at` method.
Please check it.

### CODE REVIEW

File: src/array.cr [hunk 1]
```diff
@@ -590,6 +590,51 @@ class Array(T)
...
+  # returning that elements.
+  # Raises `IndexError` if the index is out of range.
+  #
+  # ```
+  # a = ["ant", "bat", "cat", "dog"]
+  # a.delete_at(1..2)  #=> ["bat", "cat"]
+  # a                  #=> ["ant", "dog"]
+  # a.delete_at(99..100)  #=> IndexError
+  # ```
+  def delete_at(range : Range)
```
Comment (jhass) [inline src/array.cr]: We restrict `[]` etc to `Int`, I feel like we should do the same here, so `Range(Int, Int)`

File: src/array.cr [hunk 2]
```diff
@@ -590,6 +590,51 @@ class Array(T)
...
+  # ```
+  # a = ["ant", "bat", "cat", "dog"]
+  # a.delete_at(1, 2)  #=> ["bat", "cat"]
+  # a                  #=> ["ant", "dog"]
+  # a.delete_at(99,1)  #=> IndexError
+  # ```
+  def delete_at(index, count)
+    val = self[index, count]
+    index = size + index if index < 0
+    return nil if index < 0
```
Comment (jhass) [inline src/array.cr]: I think I miss a spec for this case :) (indirectly via the range overload doesn't count ;) )

File: src/array.cr [hunk 3]
```diff
@@ -590,6 +590,51 @@ class Array(T)
...
+    return [] of T if count == 0
+    count = index + count <= size ? count : size - index
```
Comment (jhass) [inline src/array.cr]: Likewise, a direct spec that hits the second branch would be great.

File: src/array.cr [hunk 4]
```diff
@@ -590,6 +590,51 @@ class Array(T)
...
+  # Raises `IndexError` if the index is out of range.
+  #
+  # ```
+  # a = ["ant", "bat", "cat", "dog"]
+  # a.delete_at(1..2)  #=> ["bat", "cat"]
+  # a                  #=> ["ant", "dog"]
+  # a.delete_at(99..100)  #=> IndexError
+  # ```
+  def delete_at(range : Range)
+    return nil if range.first < 0 && range.end > 0
```
Comment (jhass) [inline src/array.cr]: Same as the last comment, spec & probably should be empty array. Unions where we don't really need one are always a bit annoying to deal with.

**Comment by tebakane** [comment]
i see. I will change `nil` to empty array.

**Comment by tebakane** [comment]
Which do you think better IndexError or empty array for the following two syntax?
```ruby
a = [1,2,3]
a.delete_at(-1..1)
a.delete_at(-5..-4)
```

**Comment by jhass** [comment]
I think `begin > end` makes no sense, so I'd go as far and say `ArgumentError` is justified, but that's just my opinion.

**Comment by jhass** [comment]
Mmh, I guess that is while we don't have #1391

**Comment by jhass** [comment]
But then we have `(2...1).to_a` empty currently, so maybe we should make it a noop here too. I'm not sure really, I think we need some more opinions :)

**Comment by asterite** [comment]
`#delete_at` should raise on index out of bounds. The behaviour should be similar to `#[]`. If in doubt, check what Ruby does. In cases where we raise an `IndexError`, Ruby returns nil:

```crystal
# Using `slice` here because it's available in Ruby
a = [1,2,3]
a.slice(-1..1) #=> [], because it's the same as slice(1, 1)
a.delete_at(-5..-4) #=> [], because -5 is out of bounds, the minimum allowable is -3
```

I related a commit with a fix to `Array#[]`. Other specs around that spec show the expected behaviour, plus the method implementation itself :-)

**Comment by tebakane** [comment]
Add type to `delete_at` args.
`delete_at(range Range(Int, Int))` and `delete_at(index Int, count Int)`
Correct specfications like `[]` methods.

**Comment by jhass** [comment]
I took the liberty to squash the commits together, merged as e5fb9aa5342d8b19b9401d8b5ec40bbc813f40e2

Thank you!

**Pull request closed without merging**